llama_index/callbacks/open_inference_callback.py

```diff
@@ -0,0 +1,315 @@
+            max_size_in_bytes=max_size_in_bytes
+        )
+        self._document_data_buffer = DataBuffer[DocumentData](
+            max_size_in_bytes=max_size_in_bytes,
+        )
+
+    def start_trace(self, trace_id: Optional[str] = None) -> None:
+        if trace_id == "query":
+            trace_start_time = datetime.now()
+            self._trace_data = TraceData()
+            self._trace_data.query_data.timestamp = trace_start_time.isoformat()
+            self._trace_data.query_data.id = _generate_random_id()
+
+    def end_trace(
+        self,
+        trace_id: Optional[str] = None,
+        trace_map: Optional[Dict[str, List[str]]] = None,
+    ) -> None:
+        if trace_id == "query":
+            self._query_data_buffer.add(self._trace_data.query_data)
+            for document_data in self._trace_data.document_datas:
+                self._document_data_buffer.add(document_data)
+            if self._callback is not None:
+                self._callback(self._query_data_buffer)
+
+    def on_event_start(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        **kwargs: Any,
+    ) -> str:
+        if payload is not None:
+            if event_type is CBEventType.QUERY:
+                query_text = payload[EventPayload.QUERY_STR]
+                self._trace_data.query_data.id = _hash_bytes_with_sha256(
+                    query_text.encode()
+                )
+                self._trace_data.query_data.query_text = query_text
+        return event_id
+
+    def on_event_end(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        **kwargs: Any,
+    ) -> None:
+        if payload is not None:
+            if event_type is CBEventType.RETRIEVE:
+                for node_with_score in payload[EventPayload.NODES]:
+                    node = node_with_score.node
+                    score = node_with_score.score
+                    self._trace_data.query_data.document_ids.append(node.id_)
+                    self._trace_data.query_data.scores.append(score)
+                    self._trace_data.document_datas.append(
+                        DocumentData(
+                            id=node.id_,
+                            document_text=node.text,
+                        )
+                    )
+            elif event_type is CBEventType.LLM:
+                self._trace_data.query_data.response_text = payload[
+                    EventPayload.RESPONSE
+                ]
+            elif event_type is CBEventType.EMBEDDING:
+                self._trace_data.query_data.query_embedding = payload[
+                    EventPayload.EMBEDDINGS
+                ][
+                    0
+                ]  # when does this list have more than one element?
+
+    @property
+    def query_dataframe(self) -> "DataFrame":
```

Comment:
:thought: this seems slightly biases to pandas - there might be a case where someone doesn't care about a pandas representation of this data.

Reply:
Agreed, we refactored to create an `as_dataframe` function to convert a buffer to a dataframe in [88381c8](https://github.com/jerryjliu/llama_index/pull/6998/commits/88381c87e17cfcf12ad7db2766c3881ccdcb1a6d).

Reply:
I think I agree. Addressed in [a57274c](https://github.com/jerryjliu/llama_index/pull/6998/commits/a57274c0633dbaf1019685241d5312fa475550aa).